python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+                    if job not in self.active_jobs and tmp_tester.isInitialized():
+                        tmp_tester.setStatus('QUEUED', tmp_tester.bucket_pending)
+                        new_tmp_jobs.append(job)
+
+            # Add new jobs outside thread lock
+            for job in new_tmp_jobs:
+                self.runnerGo(job)
+
+        # Test is Pending and could not run for some reason or another
+        else:
+            # The runner when performing checkRunnableBase, didn't fail or skip the test. But it couldn't
+            # run for other reasons (not enough slots, etc). So place it back in the runner queue
+            if tester_name not in self.active_jobs:
+                self.runnerGo(tester_name)
+
+            # We should spin here for just a bit; A runner and status thread will rapidly hand off a
+            # job to each other for tests that can't run yet due to not enough slots being available
+            #sleep(0.01)
+
+    # Runner processing (call derived run method using a thread)
+    def jobRunner(self, queue):
+        # Get a job
+        tester_name = queue.get()
+
+        # Get the tester object based on tester_name
+        tester = self.scheduled_name_to_object[tester_name]
+
+        # Start the clock immediately (in case a derived scheduler plugin method fails to set this) or
+        # in case we attempt to check the status on this job before hitting the run method (skipped tests)
+        tester.setStartTime(clock())
+
+        can_run = True
+        with self.thread_lock:
+            # Check if we have available slots
+            if self.checkAvailableSlots(tester):
+                # Add this job to our active list
+                self.active_jobs.add(tester_name)
+
+                # Record the amount of slots we're about to consume
+                self.workers_in_use += tester.getProcs(self.options)
+
+            # We can not run. And we need to perform the status queue operation outside of this thread lock
+            else:
+                can_run = False
+
+        if can_run:
+            # Call derived run method outside of the thread_lock (run is blocking)
+            self.run(tester, self.thread_lock, self.options)
+
+            # run should never set a pending status. If it did, something is wrong with the derived scheduler
+            # TODO: create a unit test for this
+            if tester.isPending():
+                raise SchedulerError('Derived Scheduler can not return a pending status!')
+
+            # Making modifications to the tester requires a thread lock
+            with self.thread_lock:
+                # Stop the clock (in case a derived scheduler plugin method fails to set this)
+                if tester.getEndTime() is None:
+                    tester.setEndTime(clock())
+
+                # Recover our available slots
+                self.workers_in_use = max(0, self.workers_in_use - tester.getProcs(self.options))
```

Comment:
why max here? It better always be zero unless you are doing something screw for oversized jobs?